heft-plugins/heft-api-extractor-plugin/src/ApiExtractorPlugin.ts

```diff
@@ -0,0 +1,246 @@
+
+export interface IApiExtractorPluginConfiguration {
+  /**
+   * If set to true, use the project's TypeScript compiler version for API Extractor's
+   * analysis. API Extractor's included TypeScript compiler can generally correctly
+   * analyze typings generated by older compilers, and referencing the project's compiler
+   * can cause issues. If issues are encountered with API Extractor's included compiler,
+   * set this option to true.
+   *
+   * This corresponds to API Extractor's `--typescript-compiler-folder` CLI option and
+   * `IExtractorInvokeOptions.typescriptCompilerFolder` API option. This option defaults to false.
+   */
+  useProjectTypescriptVersion?: boolean;
+}
+
+export default class ApiExtractorPlugin implements IHeftTaskPlugin {
+  private _apiExtractorPromise: Promise<typeof TApiExtractor | undefined> | undefined;
+  private _apiExtractorConfigurationFilePathPromise: Promise<string | undefined> | undefined;
+  private _apiExtractorTaskConfigurationPromise:
+    | Promise<IApiExtractorPluginConfiguration | undefined>
+    | undefined;
+
+  public apply(taskSession: IHeftTaskSession, heftConfiguration: HeftConfiguration): void {
+    taskSession.hooks.clean.tapPromise(PLUGIN_NAME, async (cleanOptions: IHeftTaskCleanHookOptions) => {
+      // Load up the configuration, but ignore if target files are missing, since we will be deleting
+      // them anyway.
+      const apiExtractorConfiguration: TApiExtractor.ExtractorConfig | undefined =
+        await this._getApiExtractorConfigurationAsync(
+          taskSession,
+          heftConfiguration,
+          /* ignoreMissingEntryPoint: */ true
+        );
+      if (apiExtractorConfiguration) {
+        await this._updateCleanOptionsAsync(cleanOptions, apiExtractorConfiguration);
+      }
+    });
+
+    taskSession.hooks.run.tapPromise(PLUGIN_NAME, async (runOptions: IHeftTaskRunHookOptions) => {
+      const apiExtractor: typeof TApiExtractor | undefined = await this._getApiExtractorAsync(
+        taskSession,
+        heftConfiguration
+      );
+      const apiExtractorConfiguration: TApiExtractor.ExtractorConfig | undefined =
+        await this._getApiExtractorConfigurationAsync(taskSession, heftConfiguration);
+      if (apiExtractor && apiExtractorConfiguration) {
+        await this._runApiExtractorAsync(
+          taskSession,
+          heftConfiguration,
+          runOptions,
+          apiExtractor,
+          apiExtractorConfiguration
+        );
+      }
+    });
+  }
+
+  private async _getApiExtractorConfigurationFilePathAsync(
+    heftConfiguration: HeftConfiguration
+  ): Promise<string | undefined> {
+    if (!this._apiExtractorConfigurationFilePathPromise) {
+      this._apiExtractorConfigurationFilePathPromise =
+        heftConfiguration.rigConfig.tryResolveConfigFilePathAsync(CONFIG_FILE_LOCATION);
+    }
+    return await this._apiExtractorConfigurationFilePathPromise;
+  }
+
+  private async _getApiExtractorConfigurationAsync(
```

Comment:
It might be better ergonomically to have this method return `{ apiExtractor, extractorConfig } | undefined`, since it can't resolve the config without apiExtractor being present